### PR TITLE
Make TS InstantiationExpr parsing more permissive

### PR DESCRIPTION
**Pull request by g-plane**
This PR allows:

```ts
const x4 = f<true>
if (true) {}
```

Previously it's an error.

References:

- https://github.com/microsoft/TypeScript/issues/48654
- https://github.com/microsoft/TypeScript/pull/48659

### CODE REVIEW

File: internal/js_parser/ts_parser.go [hunk 1]
```diff
@@ -866,34 +866,36 @@ func (p *parser) canFollowTypeArgumentsInExpression() bool {
...
+		js_lexer.TStringLiteral,
+		js_lexer.TOpenBracket,
+		js_lexer.TOpenBrace,
+		js_lexer.TFunction,
+		js_lexer.TClass,
+		js_lexer.TNew,
+		js_lexer.TSlash,
+		js_lexer.TSlashEquals,
+		js_lexer.TIdentifier:
+		return false
```
Comment (evanw) [inline internal/js_parser/ts_parser.go:895]: Where did these come from? This doesn't appear to follow what Microsoft changed in their parser. For example, consider the following code:

```js
f<T>
import "x"

f<T>
import("x")
```

After their PR, [TypeScript parses it like this](https://www.typescriptlang.org/play?ts=4.7.0-dev.20220418#code/GYHgKgfAUAlgtgBwPYCcAuACARADy1KUSWRVNAClywEoog):

```js
f;
import "x";

f < T > import("x");
```

But with your PR, esbuild parses it like this:

```js
f;
import "x";
f;
import("x");
```

**Review by evanw — COMMENTED** [review]
Thanks for giving me a heads up about this change in TypeScript!

File: internal/js_parser/ts_parser.go (hunk 1 above)
Comment (g-plane) [reply internal/js_parser/ts_parser.go:895]: I forgot to handle `import` token, and I will add it.

They changed to use an existing function [`isStartOfExpression`](https://github.com/microsoft/TypeScript/blob/56a4a93718ce016f038bcbce425bd3f47c3bec78/src/compiler/parser.ts#L4235).

Comment (g-plane) [reply internal/js_parser/ts_parser.go:895]: How can I do something similar to `lookAhead` in esbuild?

Comment (evanw) [reply internal/js_parser/ts_parser.go:895]: It's kind of hacky. I stash a copy of `p.lexer` on the stack and then overwrite it again after backtracking has finished. I can take this PR from here and add the backtracking.

Comment (g-plane) [reply internal/js_parser/ts_parser.go:895]: To be honest, I hope we can have a more common method to do backtracking.

**Pull request merged**